session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/ViewTreeSnapshot.swift

```diff
@@ -13,36 +15,175 @@ import Foundation
+    /// Example: A can be not visible if it has `.zero` size or it is fully transparent.
+    let isVisible: Bool
+
+    /// If the view has any visible appearance (considering: background color + border style)
+    ///
+    /// Example: A view might have no appearance if it has `0` border width and transparent fill color.
+    let hasAnyAppearance: Bool
+}
+
+extension ViewAttributes {
+    init(frameInRootView: CGRect, view: UIView) {
+        self.frame = frameInRootView
+        self.backgroundColor = view.backgroundColor?.cgColor
+        self.layerBorderColor = view.layer.borderColor
+        self.layerBorderWidth = view.layer.borderWidth
+        self.layerCornerRadius = view.layer.cornerRadius
+        self.intrinsicContentSize = view.intrinsicContentSize
+        self.alpha = view.alpha
+
+        let hasBorderAppearance: Bool = {
+            guard view.layer.borderWidth > 0, let borderAlpha = view.layer.borderColor?.alpha else {
+                return false
+            }
+            return borderAlpha > 0
+        }()
+
+        let hasFillAppearance: Bool = {
+            guard let fillAlpha = view.backgroundColor?.cgColor.alpha else {
+                return false
+            }
+            return fillAlpha > 0
+        }()
+
+        self.isVisible = !view.isHidden && view.alpha > 0 && frame != .zero
+        self.hasAnyAppearance = self.isVisible && (hasBorderAppearance || hasFillAppearance)
     }
 }
+
+/// A type denoting semantics of given UI element in Session Replay.
+///
+/// The `NodeSemantics` is attached to each node produced by `Recorder`. During tree traversal,
+/// views are queried in available node recorders. Each `NodeRecorder` inspects the view object and
+/// tries to infer its identity (a `NodeSemantics`).
+///
+/// There are two `NodeSemantics` that describe the identity of UI element:
+/// - `AmbiguousElement` - element is of `UIView` class and we only know its base attributes (the real identity could be ambiguous);
+/// - `SpecificElement` - element is one of `UIView` subclasses and we know its specific identity along with set of subclass-specific
+/// attributes (e.g. text in `UILabel` or the "on" / "off" state of `UISwitch` control).
+///
+/// Additionally, there are two utility semantics that control the processing of nodes in SR:
+/// - `InvisibleElement` - element is either `UIView` or one of its known subclasses, but it has no visual appearance in SR, so it can
+/// be safely ignored in `Recorder` or `Processor` (e.g. a `UILabel` with no text, no border and fully transparent color).
+/// - `UnknownElement` - the element is of unknown kind, which could indicate an error during view tree traversal (e.g. working on
+/// assumption that is not met).
+///
+/// Both `AmbiguousElement` and `SpecificElement` provide an implementation of `NodeWireframesBuilder` which describes
+/// how to construct SR wireframes for UI elements they refer to. No builder is provided for `InvisibleElement` and `UnknownElement`.
```

Comment:
Because of lack of `NodeWireframesBuilder` in "unknown" and "invisible" semantics, earlier I tried to model `NodeSemantics` as enum:
```swift
enum NodeSemantics {
   case unknownElement
   case invisibleElement
   case ambiguousElement(wireframesBuilder: NodeWireframesBuilder)
   case specificElement(wireframesBuilder: NodeWireframesBuilder)
}
```
While it looked better here in declaration, it was very cumbersome for use both in `Recorder` and `Processor` (flattening part). It required `guard let case` pattern matching and was forcing numerous code branches.

I found the "`protocol` + optionals" approach to be more efficient in this kind of problem.